File: galleries/07-callbacks-both-ways/src/main.js
// 07. 함수를 양쪽으로 넘기기
//
// 06번은 파이썬이 화면을 만지는 쪽이었다. 여기서는 양쪽이 서로를 부른다.
// 새로 배우는 것은 셋이다.
//   create_once_callable 은 한 번만 부를 수 있다
//   자바스크립트에서 파이썬 함수에 키워드 인자를 넘기려면 callKwargs
//   파이썬의 키워드 인자는 자바스크립트에서 객체 하나로 묶인다

import { ensureSupport } from '../../_shared/support.js';
import { getPyodide, renderPythonError, showLoading } from '../../_shared/pyodide.js';
import { addMetricRow } from '../../_shared/metrics.js';

const bootBox = document.querySelector('#boot');
const kindFieldset = document.querySelector('#kind');
const makeButton = document.querySelector('#make');
const callButton = document.querySelector('#call');
const callKwButton = document.querySelector('#call-kw');
const releaseButton = document.querySelector('#release');
const returnButton = document.querySelector('#return');
const callLog = document.querySelector('#call-log');
const heldTable = document.querySelector('#held tbody');
const toJsButton = document.querySelector('#to-js');
const toJsResult = document.querySelector('#to-js-result');
const channelsButton = document.querySelector('#channels');
const channelsResult = document.querySelector('#channels-result');

// 파이썬이 맡긴 손잡이. 자바스크립트가 이걸 붙잡고 있다가 나중에 부른다.
// 06번의 addEventListener 가 하는 일과 같다.
let handle = null;
// 마지막으로 불러 본 결과. 참조가 있다고 부를 수 있는 것은 아니라서 따로 기억한다.
let lastCall = '아직 안 불러 봄';
let pyGlobals = null;

if (ensureSupport()) {
  start().catch((error) => renderPythonError(bootBox, error));
}

async function start() {
  bootBox.replaceChildren();
  const done = showLoading(bootBox);

  let pyodide;
  try {
    pyodide = await getPyodide();
  } catch (error) {
    done();
    renderPythonError(bootBox, error);
    return;
  }

  // 파이썬이 부를 자바스크립트 함수를 모듈로 등록한다. js 의 속성이 아니라 최상위 모듈이 된다.
  pyodide.registerJsModule('report', { name: 'report 모듈', describe, remember });

  const source = await fetch('src/main.py').then((response) => response.text());
  // 이 손잡이는 페이지가 사는 동안 계속 쓰므로 일부러 안 놓는다.
  pyGlobals = pyodide.runPython(`${source}\n\nglobals()`);
  pyGlobals.set('FROM_SET', 'globals.set 으로 넣은 값');
  done();
  bootBox.textContent = '준비됐습니다. 손잡이를 만들고 불러 보세요.';

  makeButton.addEventListener('click', () => makeHandle());
  callButton.addEventListener('click', () => callHandle(false));
  callKwButton.addEventListener('click', () => callHandle(true));
  returnButton.addEventListener('click', () => callReturned());
  releaseButton.addEventListener('click', () => releaseAll());
  toJsButton.addEventListener('click', () => showToJs());
  channelsButton.addEventListener('click', () => showChannels());

  for (const button of [makeButton, returnButton, toJsButton, channelsButton]) {
    button.disabled = false;
  }
  showHeld();
}

/** 파이썬이 손잡이를 맡기는 자리. 받아서 들고 있기만 한다. */
function remember(fn) {
  handle = fn;
}

/** 파이썬이 부를 자바스크립트 함수. 받은 것을 그대로 되읊어 준다. */
function describe(name, ...rest) {
  const shown = rest.map((value) => {
    if (value && typeof value === 'object') {
      // 진짜 객체인지 손잡이인지는 이 태그가 가른다. 이게 이 예제의 핵심 증거다.
      return `${Object.prototype.toString.call(value)} ${JSON.stringify(value)}`;
    }
    return JSON.stringify(value);
  });
  return `이름=${name} / 나머지 ${rest.length}개: ${shown.join(', ') || '(없음)'}`;
}

function currentKind() {
  return kindFieldset.querySelector('input:checked').value;
}

function call(name, ...args) {
  const fn = pyGlobals.get(name);
  try {
    return fn(...args);
  } finally {
    fn.destroy();
  }
}

function makeHandle() {
  handle = null;
  lastCall = '아직 안 불러 봄';
  call('hand_over', currentKind());
  callLog.textContent = `${currentKind()} 방식으로 손잡이를 맡았습니다. 이제 눌러 보세요.`;
  callButton.disabled = false;
  callKwButton.disabled = false;
  showHeld();
}

function callHandle(withKeywords) {
  let line;
  try {
    // 파이썬 함수에 키워드 인자를 넘기려면 callKwargs 를 쓴다. 그냥 부르면 위치 인자다.
    line = withKeywords
      ? handle.callKwargs('마루', { greeting: '반가워', excited: true })
      : handle('마루');
    lastCall = '부를 수 있음';
  } catch (error) {
    line = `실패 — ${error.message.split('\n')[0]}`;
    lastCall = '이미 죽었음';
  }

  callLog.textContent += `\n${withKeywords ? 'callKwargs' : '그냥 부르기'} → ${line}`;
  showHeld();
}

/**
 * 파이썬이 모아 둔 프록시만 놓는다.
 *
 * 자바스크립트가 든 참조는 건드리지 않는다. 모아 둔 것과 지금 든 것이 다를 수 있어서,
 * 여기서 함께 지우면 살아 있는 손잡이가 화면에서만 사라진다.
 */
function releaseAll() {
  callLog.textContent += `\n${call('release_all')}`;
  showHeld();
}

/**
 * 같은 함수를 반환값으로 받아 부른다.
 *
 * 인자로 넘긴 것과 달리 이건 빌린 손잡이가 아니다. 자동으로 회수되지 않으므로
 * 여러 번 불러도 되고, 대신 다 쓰면 우리가 놓아야 한다.
 */
function callReturned() {
  const returned = call('return_handle');
  try {
    callLog.textContent += `\n반환값으로 받아 부르기 → ${returned('마루')}`;
    callLog.textContent += `\n한 번 더 → ${returned('마루')}`;
  } finally {
    returned.destroy();
  }
}

function showHeld() {
  const held = call('held_count');
  heldTable.replaceChildren();
  addMetricRow(heldTable, '파이썬이 모아 둔 프록시', `${held}개`);
  addMetricRow(heldTable, 'greet 을 가리키는 참조 수', String(call('greet_refcount')));
  addMetricRow(heldTable, '자바스크립트가 든 참조', handle ? `있음 (${lastCall})` : '없음');
  releaseButton.disabled = held === 0;
}

function showToJs() {
  try {
    toJsResult.textContent = report(
      `위치 인자로: ${call('call_js_positional')}`,
      `키워드 인자로: ${call('call_js_keyword')}`,
      '',
      `키워드 안에 dict: ${call('nested_dict_as_keyword')}`,
      `키워드 안에 to_js: ${call('nested_to_js_as_keyword')}`,
    );
  } catch (error) {
    toJsResult.replaceChildren();
    renderPythonError(toJsResult, error);
  }
}

function report(...lines) {
  return lines.join('\n');
}

function showChannels() {
  try {
    channelsResult.textContent = call('channel_report');
  } catch (error) {
    channelsResult.replaceChildren();
    renderPythonError(channelsResult, error);
  }
}

File: galleries/07-callbacks-both-ways/src/main.py
"""양쪽으로 함수를 넘겨 본다.

report 는 자바스크립트 쪽에서 registerJsModule 로 만들어 준 모듈이다.
js 의 속성이 아니라 최상위 모듈이라서 `import report` 로 가져온다.
"""

import sys
from collections.abc import Callable
from typing import Any

import js
import report
from pyodide.ffi import JsException, create_once_callable, create_proxy, to_js

# 만들어 둔 프록시. 놓아 주려면 어딘가에 담아 둬야 한다.
# once 로 만든 것도 담는다. 부르면 알아서 놓이지만, 부르지 않고 버리면 GC 가 걷어 갈
# 때까지 남는다. 결정적으로 놓으려면 우리가 들고 있어야 한다.
_made: list[Any] = []


def greet(name: str, greeting: str = "안녕", excited: bool = False) -> str:
    """자바스크립트가 부를 파이썬 함수. 기본값 있는 인자가 있다."""
    mark = "!" if excited else "."
    return f"{greeting}, {name}{mark}"


def hand_over(kind: str) -> None:
    """세 가지 방법으로 손잡이를 만들어 자바스크립트에게 맡긴다.

    반환하지 않고 인자로 넘기는 것이 중요하다. 인자로 넘긴 파이썬 객체는 그 호출이
    끝나면 자동으로 회수되는 빌린 손잡이다. 반환값으로 넘기면 그렇지 않다.
    """
    if kind == "plain":
        # 감싸지 않고 그냥 넘긴다. report.remember 가 끝나는 순간 회수된다.
        report.remember(greet)
        return

    handle = create_once_callable(greet) if kind == "once" else create_proxy(greet)
    _made.append(handle)
    report.remember(handle)


def return_handle() -> Callable[..., str]:
    """같은 함수를 반환값으로 넘긴다. 이건 빌린 것이 아니라 넘겨준 것이다."""
    return greet


def release_all() -> str:
    """모아 둔 프록시를 전부 놓는다.

    once 는 이미 불렸으면 그때 스스로 놓였다. 그런 것에 다시 destroy 를 부르면
    던지므로 세지 않고 넘긴다.
    """
    released = 0
    already = 0
    for handle in _made:
        try:
            handle.destroy()
            released += 1
        except Exception:
            already += 1
    _made.clear()
    return f"{released}개를 놓았고 {already}개는 이미 놓여 있었습니다"


def greet_refcount() -> int:
    """greet 을 가리키는 참조 수. 손잡이가 하나 붙으면 하나 는다."""
    return sys.getrefcount(greet)


def held_count() -> int:
    return len(_made)


def call_js_positional() -> str:
    """자바스크립트 함수를 위치 인자로 부른다."""
    return report.describe("가", "여어")


def call_js_keyword() -> str:
    """키워드 인자로 부른다. 자바스크립트에는 키워드 인자가 없으므로 무언가로 바뀐다."""
    return report.describe("나", greeting="여어", excited=True)


def nested_dict_as_keyword() -> str:
    """키워드로 넘긴 값 안에 dict 가 있으면 어떻게 되는지."""
    try:
        # 보여 주려는 것은 이 한 줄이다. 나머지를 함께 감싸면 오타까지 실패로 삼킨다.
        request = js.Request.new("/x", method="POST", headers={"X": "1"})
    except JsException as exc:
        return f"실패 — {str(exc).splitlines()[0]}"
    return f"성공: {request.method}"


def nested_to_js_as_keyword() -> str:
    """같은 것을 to_js 로 감싸 넘기면."""
    try:
        request = js.Request.new("/x", method="POST", headers=to_js({"X": "1"}))
    except JsException as exc:
        return f"실패 — {str(exc).splitlines()[0]}"
    return f"성공: {request.headers.get('X')}"


def channel_report() -> str:
    """넘겨 둔 값을 어디서 꺼내는지. registerJsModule 이 특히 헷갈린다."""
    lines: list[str] = []

    def check(label: str, make: Callable[[], object]) -> None:
        try:
            lines.append(f"{label}: {make()}")
        except Exception as exc:
            lines.append(
                f"{label}: {type(exc).__name__} — {str(exc).splitlines()[0][:60]}"
            )

    # globals.set 으로 넣은 것은 파이썬 전역에 그대로 있다.
    check("globals.set 으로 넣은 것", lambda: globals().get("FROM_SET", "(없음)"))
    # registerJsModule 은 최상위 모듈이 된다. js 의 속성이 아니다.
    check("js.report 로 꺼내기", lambda: js.report.name)
    check("import report 로 꺼내기", lambda: report.name)
    return "\n".join(lines)
